### PR TITLE
Add deep reinforcement learning examples

### PR DESCRIPTION
**Pull request by muupan**
This PR will add deep reinforcement learning examples into `examples/reinforcement_learning`:

- `dqn_cartpole.py` implements DQN and DoubleDQN
- `ddpg_pendulum.py` implements DDPG

Both of the examples use OpenAI Gym as RL environments, so you need to install `gym` first.

`dqn_cartpole.py`'s 125th episode on CartPole-v0:
dqn 125th episode

`ddpg_pendulum.py`'s 64th episode on Pendulum-v0:
ddpg 64th episode

### CODE REVIEW

**Comment by rezoo** [comment]
Good PR, but I think `drl` is not intuitive, and it is better to rename `drl` to an unabbreviated form (e.g., `reinforcement_learning`).
I also think that it would be great that `dqn.py` and `ddpg.py` are renamed to `dqn_cartpole.py` and `ddpg_pendulum.py`, respectively.

**Comment by muupan** [comment]
Thanks for the review. I renamed the files accordingly.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    # Initialize a model and its optimizer
+    Q = QFunction(ndim_obs, n_actions, n_units=args.unit)
+    if args.gpu >= 0:
+        chainer.cuda.get_device(args.gpu).use()
+        Q.to_gpu(args.gpu)
+    target_Q = copy.deepcopy(Q)
+    opt = optimizers.Adam(eps=1e-2)
+    opt.setup(Q)
+
+    for episode in range(args.episodes):
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: I wonder - after the agent has mastered the task of balancing the pendulum (which often happens at episode 63 - curious why), is there a way to detect this, as further episodes won't be productive? Also, as the transition space D grows, it takes longer to run each successive episode...

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    args = parser.parse_args()
+
+    # Initialize an environment
+    env = gym.make(args.env)
+    assert isinstance(env.observation_space, gym.spaces.Box)
+    assert isinstance(env.action_space, gym.spaces.Box)
+    ndim_obs = env.observation_space.low.size
+    ndim_action = env.action_space.low.size
+    if args.record:
+        env.monitor.start(args.out, force=True)
```
Comment (zori) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Due to deprecation of env.monitor, this has to be changed to `env = gym.wrappers.Monitor(env, args.out, force=True)`. I checked and it works.

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            step += 1
+            t += 1
+
+        Rs.append(R)
+        average_R = np.mean(Rs)
+        print('episode: {} step: {} R:{} average_R:{}'.format(
+              episode, step, R, average_R))
+
+    if args.record:
+        env.monitor.close()
```
Comment (zori) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Due to deprecation of env.monitor, this is now not needed.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    args = parser.parse_args()
+
+    # Initialize an environment
+    env = gym.make(args.env)
+    assert isinstance(env.observation_space, gym.spaces.Box)
+    assert isinstance(env.action_space, gym.spaces.Discrete)
+    ndim_obs = env.observation_space.low.size
+    n_actions = env.action_space.n
+    if args.record:
+        env.monitor.start(args.out, force=True)
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. Due to deprecation of env.monitor, this has to be changed to `env = gym.wrappers.Monitor(env, args.out, force=True)`. I checked and it works.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            step += 1
+            t += 1
+
+        Rs.append(R)
+        average_R = np.mean(Rs)
+        print('episode: {} step: {} R:{} average_R:{}'.format(
+              episode, step, R, average_R))
+
+    if args.record:
+        env.monitor.close()
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. Due to deprecation of env.monitor, this is now not needed.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            new_obs, r, done, _ = env.step(a)
+            R += r
+
+            # Store a transition
+            D.append((obs, a, r * args.reward_scale, done, new_obs))
+            obs = new_obs
+
+            # Sample a random minibatch of transitions and replay
+            if len(D) >= args.replay_start_size:
+                samples = random.sample(D, args.batchsize)
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. random.sample requires "Population must be a sequence or set". Can you use a list for D instead of deque? If it's important to limit the size of the transition containers, then please cast here D to a list or set.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+def mean_clipped_loss(y, t):
+    # Add an axis because F.huber_loss only accepts arrays with ndim >= 2
+    y = F.expand_dims(y, axis=-1)
+    t = F.expand_dims(t, axis=-1)
+    return F.sum(F.huber_loss(y, t, 1.0)) / y.shape[0]
+
+
+def update(Q, target_Q, opt, samples, gamma=0.99, target_type='double_dqn'):
+    xp = Q.xp
+    s = xp.asarray([sample[0] for sample in samples], dtype=np.float32)
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. Can you rename the variables:
s -> state
a -> action
r -> reward
s_next -> state_next

Alternatively, if the letters are very common in RL literature, could you describe in the beginning of update() that samples is a list of tuples of (state, action... )?

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def mean_clipped_loss(y, t):
+    # Add an axis because F.huber_loss only accepts arrays with ndim >= 2
+    y = F.expand_dims(y, axis=-1)
+    t = F.expand_dims(t, axis=-1)
+    return F.sum(F.huber_loss(y, t, 1.0)) / y.shape[0]
+
+
+def update(Q, target_Q, opt, samples, gamma=0.99, target_type='double_dqn'):
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: The default target_type here is different than the parser's argument default.

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+def get_greedy_action(policy, obs):
+    xp = policy.xp
+    obs = xp.expand_dims(xp.asarray(obs, dtype=np.float32), 0)
+    with chainer.no_backprop_mode():
+        a = policy(obs).data[0]
+    return chainer.cuda.to_cpu(a)
+
+
+def update(Q, target_Q, policy, target_policy, opt_Q, opt_policy,
+           samples, gamma=0.99, target_type='double_dqn'):
```
Comment (zori) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Here no `target_type` needed.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    target_Q = copy.deepcopy(Q)
+    opt = optimizers.Adam(eps=1e-2)
+    opt.setup(Q)
+
+    for episode in range(args.episodes):
+
+        obs = env.reset()
+        done = False
+        R = 0.0
+        t = 0
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. t -> timestep ?

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            epsilon = 1.0 if len(D) < args.replay_start_size else \
+                max(args.min_epsilon,
+                    np.interp(
+                        step,
+                        [0, args.steps_to_decay_epsilon],
+                        [1.0, args.min_epsilon]))
+
+            # Select an action epsilon-greedily
+            if np.random.rand() < epsilon:
+                a = env.action_space.sample()
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: a -> action

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            else:
+                a = get_greedy_action(Q, obs)
+
+            # Execute an action
+            new_obs, r, done, _ = env.step(a)
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: r -> reward

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    for s, t in zip(source.params(), target.params()):
+        t.data[:] += tau * (s.data - t.data)
+
+
+def main():
+
+    parser = argparse.ArgumentParser(description='Chainer example: DRL(DDPG)')
+    parser.add_argument('--env', type=str, default='Pendulum-v0',
+                        help='Name of the OpenAI Gym environment to play')
+    parser.add_argument('--batchsize', '-b', type=int, default=64,
```
Comment (zori) [inline examples/reinforcement_learning/ddpg_pendulum.py]: batchsize -> batch-size to be same as other multi-word input arguments

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    loss.backward()
+    opt.update()
+
+
+def main():
+
+    parser = argparse.ArgumentParser(description='Chainer example: DRL(DQN)')
+    parser.add_argument('--env', type=str, default='CartPole-v0',
+                        help='Name of the OpenAI Gym environment to play')
+    parser.add_argument('--batchsize', '-b', type=int, default=64,
```
Comment (zori) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. batchsize -> batch-size to be same as other multi-word input arguments

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+    def __call__(self, s, a):
+        """Compute Q-values for given state-action pairs."""
+        x = F.concat((s, a), axis=1)
+        h = F.relu(self.l0(x))
+        h = F.relu(self.l1(h))
+        return self.l2(h)
+
+
+def squash(x, low, high):
+    """Squash values to fit [low, high] via tanh."""
```
Comment (muupan) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Yes,  the original paper of DDPG used tanh.

**Comment by ir5** [comment]
@muupan Seems that this PR is not updated for a few months, though Zori put several comments. Do you have time to update these?

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+    center = (high + low) / 2
+    scale = (high - low) / 2
+    return F.tanh(x) * scale + center
```
Comment (muupan) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Since `scale` is already divided by 2, I think it is correct.

File: examples/reinforcement_learning/dqn_cartpole.py (hunk 1 above)
Comment (muupan) [inline examples/reinforcement_learning/dqn_cartpole.py]: Since some envs defines the reward threshold at which they are considered solved, I added code to check it.

The size of D doesn't affect the time spent on each episode, so I think it doesn't matter.

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        loss = - F.sum(Q(s, policy(s))) / n
+        policy.cleargrads()
+        loss.backward()
+        opt_policy.update()
+
+    update_Q()
+    update_policy()
+
+
+def soft_copy_params(source, target, tau):
```
Comment (muupan) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Added a docstring to explain this.

**Comment by muupan** [comment]
@zori @ir5 Sorry for the delayed response. I fixed the commented points and updated the code so that it works with v2. Can you check again?

**Comment by ir5** [comment]
Great! I'll check quickly.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+    parser.add_argument('--unit', '-u', type=int, default=100,
+                        help='Number of units')
+    parser.add_argument('--target-type', type=str, default='dqn',
+                        help='Target type', choices=['dqn', 'double_dqn'])
+    parser.add_argument('--reward-scale', type=float, default=1e-2,
+                        help='Reward scale factor')
+    parser.add_argument('--replay-start-size', type=int, default=500,
+                        help=('Number of iterations after which replay is '
+                              'started'))
+    parser.add_argument('--iterations-to-decay-epsilon', type=int, default=5000,
```
Comment (ir5) [inline examples/reinforcement_learning/dqn_cartpole.py]: flake8 says "E501 line too long (80 > 79 characters)".

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+    if reward_threshold is not None:
+        print('{} defines "solving" as getting average reward of {} over 100 '
+              'consecutive trials.'.format(args.env, reward_threshold))
+    else:
+        print('{} is an unsolved environment, which means it does not have a '
+              'specified reward threshold at which it\'s considered '
+              'solved.'.format(args.env))
+
+    # Initialize variables
+    D = collections.deque(maxlen=10 ** 6)
```
Comment (ir5) [inline examples/reinforcement_learning/dqn_cartpole.py]: It might be nice if there is one line comment that explains this variable. (D is a replay buffer?)

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+    Rs = collections.deque(maxlen=100)
```
Comment (ir5) [inline examples/reinforcement_learning/dqn_cartpole.py]: Ditto. (Rs is a history of returns?)

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+        Q.to_gpu(args.gpu)
+    target_Q = copy.deepcopy(Q)
+    opt = optimizers.Adam(eps=1e-2)
+    opt.setup(Q)
+
+    for episode in range(args.episodes):
+
+        obs = env.reset()
+        done = False
+        R = 0.0
```
Comment (ir5) [inline examples/reinforcement_learning/dqn_cartpole.py]: ditto (R is a return)

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+    if reward_threshold is not None:
+        print('{} defines "solving" as getting average reward of {} over 100 '
+              'consecutive trials.'.format(args.env, reward_threshold))
+    else:
+        print('{} is an unsolved environment, which means it does not have a '
+              'specified reward threshold at which it\'s considered '
+              'solved.'.format(args.env))
+
+    # Initialize variables
+    D = collections.deque(maxlen=10 ** 6)
```
Comment (ir5) [inline examples/reinforcement_learning/ddpg_pendulum.py]: Similar with dqn example, it might be nice if there is one line comment that explains this variable. (D is a replay buffer)

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+    Rs = collections.deque(maxlen=100)
```
Comment (ir5) [inline examples/reinforcement_learning/ddpg_pendulum.py]: ditto

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+    opt_Q = optimizers.Adam()
+    opt_Q.setup(Q)
+    opt_policy = optimizers.Adam(alpha=1e-4)
+    opt_policy.setup(policy)
+
+    for episode in range(args.episodes):
+
+        obs = env.reset()
+        done = False
+        R = 0.0
```
Comment (ir5) [inline examples/reinforcement_learning/ddpg_pendulum.py]: ditto

File: examples/reinforcement_learning/ddpg_pendulum.py [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+    target_params = [param for _, param in sorted(target.namedparams())]
+    for s, t in zip(source_params, target_params):
+        t.data[:] += tau * (s.data - t.data)
+
+
+def main():
+
+    parser = argparse.ArgumentParser(description='Chainer example: DDPG')
+    parser.add_argument('--env', type=str, default='Pendulum-v0',
+                        help='Name of the OpenAI Gym environment to play')
```
Comment (ir5) [inline examples/reinforcement_learning/ddpg_pendulum.py]: "play" -> "interact", "work on" sounds more formal at least to me. Simply removing "to play" is also admissible.

File: examples/reinforcement_learning/dqn_cartpole.py [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+    Q.cleargrads()
+    loss.backward()
+    opt.update()
+
+
+def main():
+
+    parser = argparse.ArgumentParser(description='Chainer example: DQN')
+    parser.add_argument('--env', type=str, default='CartPole-v0',
+                        help='Name of the OpenAI Gym environment to play')
```
Comment (ir5) [inline examples/reinforcement_learning/dqn_cartpole.py]: ditto

**Comment by muupan** [comment]
@ir5 Thank you for the review! I fixed them.

**Comment by ir5** [comment]
Thank you. Looks good to me.
